core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/controller/promisehandlers/SkewDetectionHandler.scala

```diff
@@ -0,0 +1,150 @@
+package edu.uci.ics.amber.engine.architecture.controller.promisehandlers
+
+import com.twitter.util.Future
+import edu.uci.ics.amber.engine.architecture.controller.ControllerAsyncRPCHandlerInitializer
+import edu.uci.ics.amber.engine.architecture.controller.promisehandlers.SkewDetectionHandler.{
+  ControllerInitiateSkewDetection,
+  getSkewedAndHelperWorkersEligibleForFirstPhase,
+  previousCallFinished
+}
+import edu.uci.ics.amber.engine.architecture.deploysemantics.layer.WorkerWorkloadInfo
+import edu.uci.ics.amber.engine.common.Constants
+import edu.uci.ics.amber.engine.common.rpc.AsyncRPCServer.ControlCommand
+import edu.uci.ics.amber.engine.common.virtualidentity.ActorVirtualIdentity
+import edu.uci.ics.texera.workflow.operators.hashJoin.HashJoinOpExecConfig
+
+import scala.collection.mutable
+import scala.collection.mutable.ArrayBuffer
+import scala.util.control.Breaks.{break, breakable}
+
+object SkewDetectionHandler {
+  var previousCallFinished = true
+
+  var skewedToHelperWorkerHistory =
```

Comment:
how is this history being used, can we add some comments?

Reply:
detailed comments have been added in PR #1414